src/functions.php

```diff
@@ -29,3 +32,68 @@ function synchronized(Mutex $mutex, callable $callback, ...$args): Promise
         }
     });
 }
+
+/**
+ * Concurrently map all iterator values using {@code $processor}.
+ *
+ * The order of the items in the resulting iterator is not guaranteed in any way.
+ *
+ * @param Iterator  $iterator Values to map.
+ * @param Semaphore $semaphore Semaphore limiting the concurrency, e.g. {@code LocalSemaphore}
+ * @param callable  $processor Processing callable, which is run as coroutine. It should not throw any errors,
+ *     otherwise the entire operation is aborted.
+ *
+ * @return Iterator Mapped values.
+ */
+function concurrentMap(Iterator $iterator, Semaphore $semaphore, callable $processor): Iterator
+{
+    return new Producer(static function (callable $emit) use ($iterator, $semaphore, $processor) {
+        $processor = coroutine($processor);
+
+        /** @var \Throwable|null $error */
+        $error = null;
+        $pending = [];
+
+        while (yield $iterator->advance()) {
+            if ($error) {
+                break;
+            }
+
+            /** @var Lock $lock */
+            $lock = yield $semaphore->acquire();
+
+            $currentElement = $iterator->getCurrent();
+
+            $promise = call(static function () use (
+                $lock,
+                $currentElement,
+                $processor,
+                $emit,
+                &$error
+            ) {
+                try {
+                    yield $emit(yield $processor($currentElement));
+                } catch (\Throwable $e) {
+                    if ($error === null) {
+                        $error = $e;
+                    }
```

Comment:
Is this necessary? I'm not sure but I think the effect might be the same if you don't catch at all (just keep the finally)

Reply:
In that case `$error` is never set and it never fails?

Reply:
I meant remove the $error variable completely too... just try if the tests pass like that, if not then I missed something

Reply:
No, as `$promise` is never yielded or errors otherwise handled.